Adding BFP16 quantization/dequantization support to OSS

marjanfariborz opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63059

Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

Commit:
Adding BFP16 quantization/dequantization support to OSS

Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

wanchaol commented on torch/csrc/distributed/c10d/quantization.cpp:
```diff
@@ -0,0 +1,76 @@
+
```
Can we group the cpp/cu related files in `torch/csrc/distributed/c10d/quantization` folder instead?

wanchaol commented on torch/csrc/distributed/c10d/quantization.cpp:
```diff
@@ -0,0 +1,76 @@
+}
+
+at::Tensor _float_to_bfloat16_cpu(const at::Tensor& input) {
+
```
Maybe add some comments about what this function is doing, what's the limitation imposed (i.e. it only supports 2d tensors now).

wanchaol commented on torch/csrc/distributed/c10d/quantization.cpp:
```diff
@@ -0,0 +1,76 @@
+
+at::Tensor _float_to_bfloat16_cpu(const at::Tensor& input) {
+
+  const auto input_sizes = input.sizes();
```
can we assert that `input_sizes()` have exactly 2 dims?

wanchaol commented on torch/csrc/distributed/c10d/quantization.cpp:
```diff
@@ -0,0 +1,76 @@
+}
+
+at::Tensor _bfloat16_to_float_cpu(const at::Tensor& input) {
+  const auto input_sizes = input.sizes();
```
same, size assertion here.

wanchaol commented on torch/csrc/distributed/c10d/quantization.h:
```diff
@@ -0,0 +1,21 @@
+
+#include <ATen/ATen.h>
+
+void FloatToBFloat16Quantized_ref(
```
we don't need to expose the `ref` functions right?

wanchaol commented on torch/csrc/distributed/c10d/quantization.h:
```diff
@@ -0,0 +1,21 @@
+    const size_t ncols,
+    float* const output);
+
+at::Tensor _float_to_bfloat16_cpu(const at::Tensor& input);
```
let's wrap the exposed functions in some namespace, probably not `at::` as we don't want to conflict with internals, maybe something like `torch::distributed::c10d::quantization`?

wanchaol commented on torch/csrc/distributed/c10d/quantization_gpu.h:
```diff
@@ -0,0 +1,9 @@
+#include <ATen/ATen.h>
+#include <vector>
+
+at::Tensor _float_to_bfloat16_gpu(const at::Tensor& input);
```
same, wrap this in some namespace, global namespace is not a good practice in C++

wanchaol commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -86,13 +99,16 @@ def auto_quantize(func, qtype, quant_loss=None):
     pass other necessary arguments and then dequantizes the output.
+
     Currently it only supports:
         . FP16 quantization method
```
nit: update the doc to reflect we added the support for bfp16, also make a note about bfp16 limitation (only supports 2d tensor for now)

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

wanchaol commented on test/distributed/algorithms/quantization/test_quantization.py:
```diff
@@ -22,6 +22,14 @@ def _build_tensor(size, value=None, dtype=torch.float, device_id=None):
         return torch.empty(size, size, size, dtype=dtype).fill_(value).cuda(device_id)
 
+def _build_2Dtensor(size, value=None, dtype=torch.float, device_id=None):
+    if value is None:
```
what's the value of this util function when this is exactly the same with `_build_tensor`? can't it merge with the `_build_tensor` by specifying the dimension size in the arg? In fact I think we should just treat `size` arg as the real argument to torch.empty? like `size` should be a List[int]

wanchaol commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1,5 +1,4 @@
 #include <torch/csrc/python_headers.h>
-
```
nit: let's leave the space separation for different type of headers

wanchaol commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -31,8 +30,10 @@
+#include <c10d/quantization/quantization_gpu.h>
 #include <torch/csrc/Exceptions.h>
 #include <torch/csrc/distributed/c10d/python_comm_hook.h>
+#include <torch/csrc/distributed/c10d/quantization/quantization.h>
```
shouldn't this be the same as `<c10d/quantization/quantization_gpu.h>`? not sure why we have to include these two in two different ways, perhaps because the build_variables?

wanchaol commented on torch/csrc/distributed/c10d/quantization/quantization_utils.h:
```diff
@@ -0,0 +1,39 @@
+  return c10::DeviceTypeName(ten.device().type());
+}
+
+inline bool torch_tensor_on_cpu_check(const at::Tensor& ten) {
```
what's the necessary part about using these APIs? shouldn't the check be just directly like `tensor.is_cuda()` or not?

wanchaol commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -38,7 +41,7 @@ def _quantize_tensor_list(tensor_list, qtype):
             f"_quantize_tensor_list expecting list of torch.Tensor as input but found {type(tensor_list)}"
         )
-    if (qtype == DQuantType.FP16):
+    if ((qtype == DQuantType.FP16) or (qtype == DQuantType.BFP16)):
```
for `_quantize_tensor_list`, should we not do the check because it's doing the same check as `_quantize_tensor`? we can just leverage `_quantize_tensor` to throw error if `qtype` is not support?

wanchaol commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -76,6 +86,9 @@ def _dequantize_tensor_list(tensor_list, qtype, quant_loss=None):
     elif (qtype == DQuantType.FP16):
         dequantized_tensor_list = [_dequantize_tensor(t, qtype) for t in tensor_list]
         return dequantized_tensor_list
+    elif (qtype == DQuantType.BFP16):
```
again, `_dequantize_tensor_list` seems could just call the `dequantize_tensor` directly, we don't need to check the qtype eligibility here.

wanchaol commented on torch/csrc/distributed/c10d/quantization/quantization_gpu.cu:
```diff
@@ -0,0 +1,148 @@
+    return output;
+  }
+
+  // TODO: replace Half by BFloat16, after BFloat16 is supported by Nvidia
```
These are already supported by nvidia I believe, did you try switch to it and see if that works?

marjanfariborz replied:
This is not supported with the collective ops. This would change the dtype to bfp16 which is not supported.

wanchaol commented on torch/csrc/distributed/c10d/quantization/quantization_gpu.cu:
```diff
@@ -0,0 +1,148 @@
+      input.data_ptr<float>(),
+      nrows,
+      ncols,
+      // TODO: replace Half by BFloat16, after BFloat16 is supported by Nvidia
```
ditto

wanchaol commented on torch/csrc/distributed/c10d/quantization/quantization_gpu.cu:
```diff
@@ -0,0 +1,148 @@
+      blockDim,
+      0,
+      at::cuda::getCurrentCUDAStream()>>>(
+      // TODO: replace Half by BFloat16, after BFloat16 is supported by Nvidia
```
ditto

wanchaol reviewed (commented):
overall, looks good, have a few comments inline. Also I noticed that the CI failed to find the symbol, maybe it's related to the build system or you are adding to the wrong section to build_variables, please fix it.

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -10,7 +10,8 @@
 
 class DQuantType(Enum):
-    FP16 = "fp16"
+    FP16 = "fp16",
```
nit, and before this PR - it looks like `DQuantType` will be a public class used to pass in the `qtype` into auto_quantize API, should we add some documentation for it (can be done separately)?

rohan-varma commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -26,6 +27,8 @@ def _quantize_tensor(tensor, qtype):
     if (qtype == DQuantType.FP16):
         return _fp32_to_fp16_with_clamp(tensor)
+    elif (qtype == DQuantType.BFP16):
+        return torch.ops.q.FloatToBfloat16Quantized(tensor)
```
nit: do we want it to be a public API? If not, should we prefix it with `_` for the time being?

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1644,6 +1647,28 @@ PyMethodDef* python_functions() {
 
+namespace quantization {
+TORCH_LIBRARY(q, m) {
+    m.def("Bfloat16QuantizedToFloat(Tensor input) -> Tensor");
```
nit: prefix these with `_` to keep it private API for now?

rohan-varma commented on torch/distributed/algorithms/quantization/quantization.py:
```diff
@@ -73,26 +78,26 @@ def _dequantize_tensor_list(tensor_list, qtype, quant_loss=None):
+
     Currently it only supports:
-        . FP16 quantization method
+        . FP16 and BFP16 quantization method
```
mention which backends support with quantization methods?

rohan-varma reviewed (commented):
Awesome, great work! Just have a few comments/questions

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

wanchaol commented on tools/build_variables.bzl:
```diff
@@ -74,6 +74,8 @@ jit_core_headers = [
     "torch/csrc/jit/frontend/schema_type_parser.h",
     "torch/csrc/jit/frontend/error_report.h",
     "torch/csrc/jit/frontend/tree.h",
+    "torch/csrc/distributed/c10d/quantization/quantization_gpu.h",
```
I don't think you need to add those headers.

wanchaol commented on tools/build_variables.bzl:
```diff
@@ -733,6 +735,8 @@ libtorch_python_distributed_core_sources = [
     "torch/csrc/distributed/c10d/init.cpp",
     "torch/csrc/distributed/c10d/python_comm_hook.cpp",
+    "torch/csrc/distributed/c10d/quantization/quantization.cpp",
+    "torch/csrc/distributed/c10d/quantization/quantization_gpu.cu",
```
let's add the cuda source file to libtorch_cuda_distributed_extra_sources

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

wanchaol commented on torch/csrc/distributed/c10d/quantization/quantization_gpu.cu:
```diff
@@ -0,0 +1,147 @@
+#include <caffe2/operators/segment_reduction_op_gpu.cuh>
```
what make us using this include? I don't see any call site calling this file

marjanfariborz replied:
Some of the functions and parameters are enabled by this (C10_CUDA_KERNEL_LAUNCH_CHECK, getCurrentCUDAStream, OptionalCUDAGuard, ...). I'm trying to find their sources and include them directly.

wanchaol replied:
All these macro/functions are not related to that op, so we shouldn't include that op and instead we should directly include those relevant headers (which I think it should resides somewhere in the torch/c10 directly or torch/cuda directory

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/quantization/quantization.cpp:
```diff
@@ -0,0 +1,93 @@
+#include <torch/csrc/distributed/c10d/quantization/quantization.h>
```
These seem like general quantization libraries, should these be in c10d or some general torch folder which has rest of quant stuff?

marjanfariborz replied:
That's good point. I initially had them in the c10d but I didn't see any other quantization code in that directory that's why I moved them all to a separate folder.

wanchaol replied:
I feel like the reason we might leave this in distributed folder is, there's a limitation different backends (MPI, Gloo) for not accepting bfp16 dtyped tensor, in torch quantization area user can do things like `to(torch.bfloat16)` and `to(torch.float32)`  to do quantization/dequant. Once we figure out the way to support bfp16 tensor in the comm backend, we can use `to` instead of custom kernels.

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

wanchaol commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -31,14 +32,21 @@
 #include <torch/custom_class.h>
 
+#include <ATen/ATen.h>
+#include <ATen/TypeDefault.h>
```
I don't think any of these four includes are needed, maybe <ATen/ATen.h> (although I don't think that's needed as well).

wanchaol commented on test/distributed/algorithms/quantization/test_quantization.py:
```diff
@@ -38,13 +39,13 @@ def _build_tensor(size, value=None, dtype=torch.float, device_id=None):
         def setUp(self):
             super(DistQuantizationTests, self).setUp()
             self._spawn_processes()
+            # WORLD_SIZE = int(os.environ["WORLD_SIZE"])
```
remove the comment?

wanchaol commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -1644,6 +1652,29 @@ PyMethodDef* python_functions() {
+    m.def("_Bfloat16QuantizedToFloat(Tensor input) -> Tensor");
+    m.def("_FloatToBfloat16Quantized(Tensor input) -> Tensor");
+}
+#ifdef USE_C10D_GLOO
```
do we need this macro?

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

wanchaol commented on test/distributed/algorithms/quantization/test_quantization.py:
```diff
@@ -117,23 +128,46 @@ def test_all_to_all_fp16(self):
-                    allgather = quant.auto_quantize(dist.all_gather, qtype, quant_loss=None)
-                else:
-                    allgather = dist.all_gather
+                tensor = _build_tensor([dest + 1, dest + 1, dest + 1], rank, dtype=dtype)
```
is it necessary to distinguish between fp16 and bfp16 when building tensor inputs? why don't we just all use 2 dim tensor?

marjanfariborz replied:
This is the datatype (dtype) I just wanted to make sure that the input tensor is FP32.

wanchaol replied:
what does it mean? I was referring to the IF branch below that you are creating a two dimensional tensor for BFP16 case, but for other cases we are using three dimensional tensor, why not we just creating two dimensional tensor for all cases?

wanchaol reviewed (approved):
lgtm, just one minor comment, please ensure CI pass when landing

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

wanchaol commented on test/distributed/algorithms/quantization/test_quantization.py:
```diff
@@ -26,7 +27,7 @@ def _build_tensor(size, value=None, dtype=torch.float, device_id=None):
-        return torch.empty(size, size, size, dtype=dtype).fill_(value)
+        return torch.empty(size, dtype=dtype).fill_(value)
     else:
         return torch.empty(size, size, size, dtype=dtype).fill_(value).cuda(device_id)
```
why we don't change the torch.empty here? we should all using `torch.empty(size, dtype=dtype)`?

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]

Commit:
Update on "Adding BFP16 quantization/dequantization support to OSS"


Supporting BFP16 quantization method to OSS. Currently only support CPU

Differential Revision: [D30194538](https://our.internmc.facebook.com/intern/diff/D30194538/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D30194538/)!

[ghstack-poisoned]
